api/rest/client/lbclient.go

```diff
@@ -0,0 +1,421 @@
+package client
+
+import (
+	"context"
+	"sync/atomic"
+
+	cid "github.com/ipfs/go-cid"
+	shell "github.com/ipfs/go-ipfs-api"
+	files "github.com/ipfs/go-ipfs-files"
+	"github.com/ipfs/ipfs-cluster/api"
+	peer "github.com/libp2p/go-libp2p-peer"
+)
+
+// loadBalancingClient is a client to interact with IPFS Cluster APIs
+// that balances the load by distributing requests among peers.
+type loadBalancingClient struct {
+	strategy LBStrategy
+	retries  int
+}
+
+// LBStrategy is a strategy to load balance requests among clients.
+type LBStrategy interface {
+	Next(count int) Client
+	SetClients(clients []Client)
+}
+
+// RoundRobin is a load balancing strategy that would use clients in a sequence.
+type RoundRobin struct {
+	clients []Client
+	counter uint32
+	length  uint32
+}
+
+// Next return the next client to be used.
+func (r *RoundRobin) Next(count int) Client {
+	i := atomic.AddUint32(&r.counter, 1) % r.length
+
+	return r.clients[i]
+}
+
+// SetClients sets a list of clients for this strategy.
+func (r *RoundRobin) SetClients(cl []Client) {
+	r.clients = cl
+}
+
+// Failover is a load balancing strategy that would try the local cluster peer first.
+// If the local call fail it would try other client in a round robin fashion.
+type Failover struct {
+	clients []Client
+	counter uint32
+	length  uint32
+}
+
+// Next returns the next client to be used.
+func (f *Failover) Next(count int) Client {
+	if count == 0 {
```

Comment:
I think this function can be summarized to `return f.clients[count]`. You are sort of load-balancing all clients except the first. Failover just means `try the next in the list`.

Reply:
> Failover just means try the next in the list

ok, I can change this. What I have for Failover, I think, should be better called RoundRobin.

Reply:
I will make RoundRobin as well, always use the first client.

Reply:
The other strategy is round robin because it's a global roundrobin. This is not a global roundrobin in the sense that clients are always used in order based on the current count of retries.

Don't touch the RoundRobin client, it does not have to always use the first client (why would it).

Reply:
> (why would it)

RoundRobin currently, for the first request try, makes a slower remote call instead of local one.

Reply:
who said the first client is a local one?